docs/tutorial-part2.rst

```diff
@@ -0,0 +1,550 @@
+Tutorial (Part 2)
+=================
+
+
+.. note::
+
+   This document continues from where the first part left off.
+   As in the first part, we proceed by building on the example application
+   step by step.
+   It would be helpful to the reader to code along
+   and try out the various stages of the application.
+   In this manner, completing it should take about an hour.
+
+
+The first part of the tutorial covered some basic topics
+like routing, form handling, static assets, and JSON endpoints.
+This second part will cover resource handling, redirection,
+and middleware usage.
+
+The example application will be a link shortener.
+There will be an option for letting shortened links expire,
+based on time or on the number of times they have been clicked.
+For the sake of simplicity, we'll use the ``shelve`` module
```

Comment:
Could you look into the `storage.py` implementation? I'm not comfortable with it at all. Opening / closing the file every time doesn't feel right. Maybe I should open once and then use `.sync()` when needed? How do I make sure to close it when the application exits or crashes?

Reply:
:mod:`shelve` didn't create a link when I tried. Maybe because the shelve module is not part of the project? Anyway, I've linked to the python documentation site.